MATLAB/Algorithms/LSQR.m

```diff
@@ -0,0 +1,273 @@
+%   using the projection data PROJ taken over ALPHA angles, corresponding
+%   to the geometry descrived in GEO, using NITER iterations.
+% 
+%  LSQR_CBCT(PROJ,GEO,ANGLES,NITER,OPT,VAL,...) uses options and values for solving. The
+%   possible options in OPT are:
+% 
+% 
+%  'Init'    Describes diferent initialization techniques.
+%             * 'none'     : Initializes the image to zeros (default)
+%             * 'FDK'      : intializes image to FDK reconstrucition
+%             * 'multigrid': Initializes image by solving the problem in
+%                            small scale and increasing it when relative
+%                            convergence is reached.
+%             * 'image'    : Initialization using a user specified
+%                            image. Not recomended unless you really
+%                            know what you are doing.
+%  'InitImg'    an image for the 'image' initialization. Avoid.
+% 'redundancy_weighting': true or false. Default is true. Applies data
+%                         redundancy weighting to projections in the update step
+%                         (relevant for offset detector geometry)
+%--------------------------------------------------------------------------
+%--------------------------------------------------------------------------
+% This file is part of the TIGRE Toolbox
+% 
+% Copyright (c) 2015, University of Bath and 
+%                     CERN-European Organization for Nuclear Research
+%                     All rights reserved.
+%
+% License:            Open Source under BSD. 
+%                     See the full license at
+%                     https://github.com/CERN/TIGRE/blob/master/LICENSE
+%
+% Contact:            tigre.toolbox@gmail.com
+% Codes:              https://github.com/CERN/TIGRE/
+% Coded by:           Ander Biguri 
+%--------------------------------------------------------------------------
+
+%%
+
+% msl: I assume this x is either an array of 0 of the appropriate
+% dimensions or an initial guess
+
+[verbose,x,QualMeasOpts,gpuids,redundancy_weights]=parse_inputs(proj,geo,angles,varargin);
+
+% msl: no idea of what this is. Should I check?
+measurequality=~isempty(QualMeasOpts);
+qualMeasOut=zeros(length(QualMeasOpts),niter);
+if redundancy_weights
+    % Data redundancy weighting, W_r implemented using Wang weighting
+    % reference: https://iopscience.iop.org/article/10.1088/1361-6560/ac16bc
+    
+    num_frames = size(proj,3);
+    W_r = redundancy_weighting(geo);
+    W_r = repmat(W_r,[1,1,num_frames]);
+    % disp('Size of redundancy weighting matrix');
+    % disp(size(W_r));
+end
+
+% Paige and Saunders //doi.org/10.1145/355984.355989
+
+% Enumeration as given in the paper for 'Algorithm LSQR'
+% (1) Initialize 
+u=proj-Ax(x,geo,angles,'Siddon','gpuids',gpuids);
+normr = norm(u(:),2);
+u = u/normr;
+
+beta = normr;
+phibar = beta;
+
+% msl: I have not checked 'redundancy_weights', do they have to do with 
+% mismatched transpose? 
+if redundancy_weights 

```

Comment:
No, we should remove them. They are weigths that minimize overlap of data when there is offset of detector. They are quite heuristic, so I think they will screw up our nice Krylov spaces.